Migrate to register(storeDescriptor) for items data store

tomalec opened the pull request:
### Submission Review Guidelines:

-   I have followed the [WooCommerce Contributing Guidelines](https://github.com/woocommerce/woocommerce/blob/trunk/.github/CONTRIBUTING.md) and the [WordPress Coding Standards](https://make.wordpress.org/core/handbook/best-practices/coding-standards/).
-   I have checked to ensure there aren't other open [Pull Requests](https://github.com/woocommerce/woocommerce/pulls) for the same update/change.
-   I have reviewed my code for [security best practices](https://developer.wordpress.org/apis/security/).
-   Following the above guidelines will result in quick merges and clear and detailed feedback when appropriate.

<!-- You can erase any parts of this template not applicable to your Pull Request. -->

### Changes proposed in this Pull Request:

<!-- If necessary, indicate if this PR is part of a bigger feature. Add a label with the format `focus: name of the feature [team:name of the team]`. -->

<!-- Describe the changes made to this Pull Request and the reason for such changes. -->

<!-- For bug fixes: If known, please provide links to help with traceability and escape analysis. -->
<!-- Please include a link to the issue of the bug being fixed, if one doesn't exist please create it. -->
<!-- If the PR that introduced the bug is known, please also add its link below. -->

Closes https://github.com/woocommerce/woocommerce/issues/55380

As part of the React 18 upgrade and WordPress 6.6 compatibility changes, this PR updates the items store registration pattern in @woocommerce/data to use the modern WordPress data store API.

Key changes include:

1. Migrated from deprecated `registerStore` to `createReduxStore` and `register` pattern
2. Updated store exports to match the modern pattern used across other stores:
	- Added `export { store as itemsStore }`
	- Maintained `ITEMS_STORE_NAME` export for backward compatibility
4. Update items store import to use `itemsStore` from `@woocommerce/data`

### Screenshots or screen recordings:

<!-- If this PR includes UI changes, please provide screenshots or a screen recording for clarity. -->
<!-- This section can be removed if not applicable. -->

| Before | After |
| ------ | ----- |
|        |       |


<!-- Begin testing instructions -->

### How to test the changes in this Pull Request:

<!-- Include detailed instructions on how these changes can be tested. Review and follow the guide for how to write high-quality testing instructions. -->

Using the [WooCommerce Testing Instructions Guide](https://github.com/woocommerce/woocommerce/wiki/Writing-high-quality-testing-instructions), include your detailed testing instructions:

This PR is a developer-facing, doesn't change any functionality.

For devs:

1. Review the `@woocommerce/data` changes for the items store migration
2. Delete the `noCheck: true` line in `./packages/js/data/tsconfig.json` and running `npx tsc` inside `./packages/js/data`
3. Confirm that there are no TS errors in `./src/items/**`
4. `window.wp.data.select( wc.data.itemsStore )` and `window.wp.data.dispatch( wc.data.itemsStore )` should return selectors and dispatchers
5. Smoke test analytics pages

<!-- End testing instructions -->

### Testing that has already taken place:

<!-- Detail any testing that has already been conducted. -->
<!-- Include environment details such as hosting type, plugins, theme, store size, store age, and relevant settings. -->
<!-- Mention any analysis performed, such as assessing potential impacts on environment attributes and other plugins, performance profiling, or LLM/AI-based analysis. -->
<!-- Within the testing details you provide, please ensure that no sensitive information (such as API keys, passwords, user data, etc.) is included in this public pull request. -->


I performed steps above on a local environment with WP 6.7.1. Asked ChatGPT for the review of changes here.

### Changelog entry

<!-- You can optionally choose to enter a changelog entry by checking the box below and supplying data. -->
<!-- It will trigger the 'Add changelog to PR' CI job to create and push the entry into the branch. -->

<!-- Due to org permissions, the job might fail for PRs crated from a fork under GitHub organizations. Possible solutions: -->
<!-- * Create entry manually with `pnpm --filter='@woocommerce/plugin-woocommerce' changelog add` and push it into the branch (replace `@woocommerce/plugin-woocommerce` with package name from nearest `package.json` file) -->
<!-- * Create entry from supplied PR data and push it automatically `pnpm utils changefile pr-number-here -o github-org-name-here` -->

-   [ ] Automatically create a changelog entry from the details below.

<!-- If no changelog entry is required for this PR, you can specify that below and provide a comment explaining why. This cannot be used if you selected the option to automatically create a changelog entry above. -->

-   [ ] This Pull Request does not require a changelog entry. (Comment required below)

<details>

<summary>Changelog Entry Details</summary>

#### Significance

<!-- Choose only one -->

-   [ ] Patch
-   [ ] Minor
-   [ ] Major

#### Type

<!-- Choose only one -->

-   [ ] Fix - Fixes an existing bug
-   [ ] Add - Adds functionality
-   [ ] Update - Update existing functionality
-   [ ] Dev - Development related task
-   [ ] Tweak - A minor adjustment to the codebase
-   [ ] Performance - Address performance issues
-   [ ] Enhancement - Improvement to existing functionality

#### Message <!-- Add a changelog message here -->

</details>

<details>

<summary>Changelog Entry Comment</summary>

#### Comment <!-- If your Pull Request doesn't require a changelog entry, a comment explaining why is required instead -->

</details>

tomalec committed:
Migrate to `register(storeDescriptor)` for items data store

Address https://github.com/woocommerce/woocommerce/issues/55380

tomalec committed:
Update items store import to use itemsStore from @woocommerce/data

Address https://github.com/woocommerce/woocommerce/issues/55380

tomalec commented:
> The unused exports and types after the migration should be removed, i.e.:
> • unused selectors type export
> • declare module @wordpress/data type override.

Reports selectors use generics not supported by `wp-data` or `WPDataSelector` type utils. I tried to make a util that would convert all selectors preserving type parameters, but I didn't find a way to do that effectively. 

So I left the type definition and the declare override as it were.

tomalec commented:
• the REPORTS_STORE_NAME string

I noticed all other stores keep their names for backward compatibility, so I did the same in https://github.com/woocommerce/woocommerce/pull/56421

tomalec committed:
Add changelog entries

tomalec committed:
Fix whitespace for eslint

eason9487 commented on plugins/woocommerce/client/admin/client/analytics/report/index.js:
```diff
@@ -113,7 +113,7 @@ export default compose(
 
 		/* eslint @wordpress/no-unused-vars-before-return: "off" */
-		const itemsSelector = select( ITEMS_STORE_NAME );
+		const itemsSelector = select( itemsStore );
```
Maybe moving this line closer to where the variable will be used to eliminate the eslint off annotation.

eason9487 commented on plugins/woocommerce/client/admin/client/dashboard/leaderboards/index.js:
```diff
@@ -180,7 +180,7 @@ Leaderboards.propTypes = {
 export default compose(
 	withSelect( ( select ) => {
-		const { getItems, getItemsError } = select( ITEMS_STORE_NAME );
+		const { getItems, getItemsError } = select( itemsStore );
```
These two selectors are passed to `Leaderboards` component but don't seem to be used. Maybe consider removing them.

eason9487 commented on plugins/woocommerce/client/admin/client/homescreen/activity-panel/orders/index.js:
```diff
@@ -275,7 +275,7 @@ function OrdersPanel( { unreadOrdersCount, orderStatuses } ) {
 			select( ordersStore );
 		// eslint-disable-next-line @wordpress/no-unused-vars-before-return
-		const { getItems } = select( ITEMS_STORE_NAME );
+		const { getItems } = select( itemsStore );
```
Maybe moving this line closer to where the variable will be used to eliminate the `eslint-disable-next-line` annotation.

eason9487 reviewed (approved):
Thanks for the migration. LGTM.

All review comments are not blockers.

tomalec committed:
Move `select( itemsStore )` closer to where it's used

Address https://github.com/woocommerce/woocommerce/pull/56424#pullrequestreview-2701464216

tomalec committed:
Merge branch 'trunk' into dev/55380-items-store

tomalec committed:
Remove store select from `Leaderboards` component

Address https://github.com/woocommerce/woocommerce/pull/56424#discussion_r2004960847

tomalec committed:
Use `select( itemsStore)` in utils.

tomalec commented:
Thank you @eason9487 for catching those!
I moved calls closer to where the results are used and reduced the `Leaderboards` implementation. I would appreciate another look, as changes in https://github.com/woocommerce/woocommerce/pull/56424/commits/8e53cda85a3e8010fa25de8af284ad512b2508a8  and https://github.com/woocommerce/woocommerce/pull/56424/commits/7d49a34cdc781b4c755cc02206338b752507dfb6 are affecting the implementation a bit more.

tomalec commented:
I'm sorry, but the unit tests are failing due to `reducer` being undefined for some reason. I guess there may be some mismatch in references/imports somewhere, as it happens only in test environments. I'll continue the investigation.

eason9487 commented:
> [...] but the unit tests are failing due to `reducer` being undefined for some reason. I guess there may be some mismatch in references/imports somewhere, as it happens only in test environments. I'll continue the investigation.

The https://github.com/woocommerce/woocommerce/pull/56424/commits/7d49a34cdc781b4c755cc02206338b752507dfb6 change forms a circular dependency under `packages/js/data/src`. The following dependency occurs when importing `items/reducer.ts`:

1. `items/reducer.ts`
1. `items/utils.ts`
1. `index.ts`
1. `items/index.ts`
1. `items/reducer.ts`

This causes `items/index.ts` to use `items/reducer.ts` before the latter module has finished initializing.

tomalec committed:
Resolve circular dependency for `getLeaderboard`

tomalec committed:
Merge branch 'trunk' into dev/55380-items-store

tomalec committed:
Merge branch 'trunk' into dev/55380-items-store

eason9487 reviewed (approved):
I did some testing on the updates and it's working fine. LGTM. Thanks!

The unrelated failed E2E tests should have been fixed on `trunk` now. So I'm approving this PR in advance.

tomalec committed:
Merge branch 'trunk' into dev/55380-items-store
